alibi_detect/utils/saving.py

```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
 
-    return od
+    return preprocess_cfg
 
 
-def init_od_s2s(state_dict: Dict,
-                seq2seq: tf.keras.Model) -> OutlierSeq2Seq:
+def _serialize_function(func: Callable, base_path: Path,
+                        local_path: Path = Path('function')) -> Tuple[str, dict]:
     """
-    Initialize OutlierSeq2Seq.
+    Serializes a generic function. If the function is in the object registry, the registry str is returned.
+    The function is saved to dill, and if wrapped in a functools.partial, the kwargs are returned.
 
     Parameters
     ----------
-    state_dict
-        Dictionary containing the parameter values.
-    seq2seq
-        Loaded seq2seq model.
+    func
+        The function to serialize.
+    base_path
+        Base directory to save in.
+    local_path
+        A local (relative) filepath to append to base_path. Default is 'function/'.
 
     Returns
     -------
-    Initialized OutlierSeq2Seq instance.
+    Tuple containing a string referencing the save filepath and a dict of kwargs.
     """
-    seq_len, n_features = state_dict['shape'][1:]
-    od = OutlierSeq2Seq(n_features,
-                        seq_len,
-                        threshold=state_dict['threshold'],
-                        seq2seq=seq2seq,
-                        latent_dim=state_dict['latent_dim'],
-                        output_activation=state_dict['output_activation'])
-    return od
-
-
-def load_text_embed(filepath: Union[str, os.PathLike], load_dir: str = 'model') \
-        -> Tuple[TransformerEmbedding, Callable]:
-    model_dir = Path(filepath).joinpath(load_dir)
-    tokenizer = AutoTokenizer.from_pretrained(str(model_dir.resolve()))
-    args = dill.load(open(model_dir.joinpath('embedding.dill'), 'rb'))
-    emb = TransformerEmbedding(
-        str(model_dir.resolve()), embedding_type=args['embedding_type'], layers=args['layers']
-    )
-    return emb, tokenizer
-
-
-def init_preprocess(state_dict: Dict, model: Optional[Union[tf.keras.Model, tf.keras.Sequential]],
-                    emb: Optional[TransformerEmbedding], tokenizer: Optional[Callable], **kwargs) \
-        -> Tuple[Optional[Callable], Optional[dict]]:
-    """ Return preprocessing function and kwargs. """
-    if kwargs:  # override defaults
-        keys = list(kwargs.keys())
-        preprocess_fn = kwargs['preprocess_fn'] if 'preprocess_fn' in keys else None
-        preprocess_kwargs = kwargs['preprocess_kwargs'] if 'preprocess_kwargs' in keys else None
-        return preprocess_fn, preprocess_kwargs
-    elif model is not None and isinstance(state_dict['preprocess_fn'], Callable) \
-            and isinstance(state_dict['preprocess_kwargs'], dict):
-        preprocess_fn = state_dict['preprocess_fn']
-        preprocess_kwargs = state_dict['preprocess_kwargs']
+    # If a partial, save function and kwargs
+    if isinstance(func, partial):
+        kwargs = func.keywords
+        func = func.func
     else:
-        return None, None
+        kwargs = {}
```

Comment:
Why do we have the two cases here? partial function and kwargs / function and no kwargs?

Reply:
P.s. leaving unresolved in case of following questions.

Reply:
ok, makes sense!